File: infer/tests/codetoanalyze/java/performance/ForEachTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import java.util.List;
import java.util.Map;

class ForEachTest {

  int add(Integer x, Integer y) {
    return x + y;
  }

  int loop_linear(Integer x, List<Integer> list) {
    int sum = 0;
    for (Integer el : list) {
      sum = +el + x;
    }
    return sum;
  }

  void map_linear(Map<Integer, Integer> map) {
    map.forEach((key, value) -> add(key, value));
  }

  void list_linear(List<Integer> myList) {
    myList.forEach(el -> add(el, 1));
  }

  // FN: We have limited lambda support and canot incur costs of the lambda calls yet.
  void list_quadratic_FN(List<Integer> myList) {
    myList.forEach(el -> loop_linear(el, myList));
  }
}

File: infer/tests/codetoanalyze/cpp/performance/foreachtest.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <algorithm>
#include <list>
#include <map>
#include <vector>
#include <iostream>

int add(int x, int y) { return x + y; }

int loop_linear_list(int x, const std::list<int>& list) {
  int sum = 0;
  for (auto&& el : list) {
    sum = +el + x;
  }
  return sum;
}

int loop_linear_vec(int x, const std::vector<int>& vec) {
  int sum = 0;
  for (auto&& el : vec) {
    sum = +el + x;
  }
  return sum;
}

void map_linear_FN(std::map<int, int>& map) {
  for_each(map.begin(), map.end(), [](std::pair<int, int> el) {
    add(el.first, el.second);
  });
}

// FN: We have limited lambda support and canot incur costs of the lambda calls
// yet.
void list_quadratic_FN(std::list<int>& mylist) {
  for_each(mylist.begin(), mylist.end(), [mylist](int el) {
    loop_linear_list(el, mylist);
  });
}

void vector_quadratic_FN(std::vector<int>& vec) {
  for_each(vec.begin(), vec.end(), [vec](int el) { loop_linear_vec(el, vec); });
}
